sentry-options-validation/src/lib.rs

```diff
@@ -0,0 +1,384 @@
+#[derive(Debug, thiserror::Error)]
+pub enum ValidationError {
+    #[error("Schema error in {file}: {message}")]
+    SchemaError { file: PathBuf, message: String },
+
+    #[error("Value error for {namespace}: {errors}")]
+    ValueError { namespace: String, errors: String },
+
+    #[error("Unknown namespace: {0}")]
+    UnknownNamespace(String),
+
+    #[error("Failed to read file: {0}")]
+    FileRead(#[from] std::io::Error),
+
+    #[error("Failed to parse JSON: {0}")]
+    JSONParse(#[from] serde_json::Error),
+}
+
+/// Schema for a namespace, containing only a validator
+pub struct NamespaceSchema {
+    pub namespace: String,
+    _validator: jsonschema::Validator,
+}
+
+impl NamespaceSchema {
+    /// Validate an entire values object against this schema
+    ///
+    /// # Arguments
+    /// * `values` - JSON object containing option key-value pairs
+    ///
+    /// # Errors
+    /// Returns error if values don't match the schema
+    pub fn validate_values(&self, _values: &Value) -> ValidationResult<()> {
+        // TODO: Implement validation
+        Ok(())
+    }
+}
+
+/// Registry for loading and storing schemas
+pub struct SchemaRegistry {
+    schemas: HashMap<String, Arc<NamespaceSchema>>,
+}
+
+impl SchemaRegistry {
+    /// Create a new empty schema registry
+    pub fn new() -> Self {
+        Self {
+            schemas: HashMap::new(),
+        }
+    }
+
+    /// Load schemas from a directory and create a registry
+    ///
+    /// Expects directory structure: `schemas/{namespace}/schema.json`
+    ///
+    /// # Arguments
+    /// * `schemas_dir` - Path to directory containing namespace subdirectories
+    ///
+    /// # Errors
+    /// Returns error if directory doesn't exist or any schema is invalid
+    pub fn from_directory(schemas_dir: &Path) -> ValidationResult<Self> {
+        let schemas = Self::load_all_schemas(schemas_dir)?;
+        Ok(Self { schemas })
+    }
+
+    /// Validate an entire values object for a namespace
+    ///
+    /// # Arguments
+    /// * `namespace` - Namespace name
+    /// * `values` - JSON object containing option key-value pairs
+    ///
+    /// # Errors
+    /// Returns error if namespace doesn't exist or values don't match schema
+    pub fn validate_values(&self, namespace: &str, values: &Value) -> ValidationResult<()> {
+        let schema = self
+            .schemas
+            .get(namespace)
+            .ok_or_else(|| ValidationError::UnknownNamespace(namespace.to_string()))?;
+
+        schema.validate_values(values)
+    }
+
+    /// Load all schemas from a directory
+    fn load_all_schemas(
+        schemas_dir: &Path,
+    ) -> ValidationResult<HashMap<String, Arc<NamespaceSchema>>> {
+        if !schemas_dir.exists() {
+            return Err(ValidationError::FileRead(std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                format!("Schemas directory not found: {}", schemas_dir.display()),
+            )));
+        }
```

Comment:
Technically not necessary since we catch this same error on line 124 when we call `fs::read_dir(schemas_dir)?`. If we want to remove an unnecessary call but keep the nice error message, we could do a map_err on the `fs::read_dir()` before looping through the entries instead.

Reply:
tru